Developing and use .ts

Hi there,

When using the new beta method, of:
`await spawn(new Worker('../controller/' + command));`
Also note that this is not allowing .ts extensions? It forces looking for .js extensions? Anyway I can bypass this?

I get this:

```
{ Error: Cannot find module '/Users/user_1/Websites/bf/Server/source/controller/auctions.js'
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:613:15)
    at Function.resolve (internal/modules/cjs/helpers.js:28:19)
    at resolveScriptPath (/Users/user_1/Websites/bf/Server/node_modules/threads/dist/master/implementation.node.js:32:19)
    at new Worker (/Users/user_1/Websites/bf/Server/node_modules/threads/dist/master/implementation.node.js:42:19)
    at ThreadManager.<anonymous> (/Users/user_1/Websites/bf/Server/source/module/threads.ts:41:41)
    at Generator.next (<anonymous>)
    at /Users/user_1/Websites/bf/Server/source/module/threads.ts:7:71
    at new Promise (<anonymous>)
    at __awaiter (/Users/user_1/Websites/bf/Server/source/module/threads.ts:3:12)
    at ThreadManager.execute (/Users/user_1/Websites/bf/Server/source/module/threads.ts:20:16) code: 'MODULE_NOT_FOUND' }
```

~ Danny

--- a/.travis.yml
+++ b/.travis.yml
@@ -16,7 +16,7 @@ matrix:
         grep: stable
       filter_secrets: false  # Secrets will otherwise break Windows builds
       node_js: 8.6
-      script: npx run-s pretest test:ava
+      script: npm run test:ava
       env:
         - THREADS_WORKER_INIT_TIMEOUT=30000
     - os: windows
@@ -24,7 +24,7 @@ matrix:
         grep: stable
       filter_secrets: false  # Secrets will otherwise break Windows builds
       node_js: latest
-      script: npx run-s pretest test:ava
+      script: npm run test:ava
 
 env:
   - THREADS_WORKER_INIT_TIMEOUT=15000

--- a/README.md
+++ b/README.md
@@ -181,6 +181,7 @@ In case of exposing an object, `spawn()` will asynchronously return an object ex
 
 - [**Introduction**](https://threads.js.org/usage#introduction)
 - [**Basics**](https://threads.js.org/usage#basics)
+- [**TypeScript**](https://threads.js.org/usage#typescript)
 - [**Observables**](https://threads.js.org/usage#observables)
 - [**Thread pool**](https://threads.js.org/usage#thread-pool)
 - [**Transferable objects**](https://threads.js.org/usage#transferable-objects)

--- a/docs/usage.md
+++ b/docs/usage.md
@@ -118,7 +118,32 @@ const counter = {
 expose(counter)
 ```
 
-### TypeScript
+### Error handling
+
+Works fully transparent - the promise in the master code's call will be rejected with the error thrown in the worker, also yielding the worker error's stack trace.
+
+```js
+// master.js
+import { spawn, Thread, Worker } from "threads"
+
+const counter = await spawn(new Worker("./workers/counter"))
+
+try {
+  await counter.increment()
+  await counter.increment()
+  await counter.decrement()
+
+  console.log(`Counter is now at ${await counter.getCount()}`)
+} catch (error) {
+  console.error("Counter thread errored:", error)
+} finally {
+  await Thread.terminate(counter)
+}
+```
+
+## TypeScript
+
+### Type-safe threads
 
 When using TypeScript you can declare the type of a `spawn()`-ed thread:
 
@@ -166,28 +191,15 @@ export type Counter = typeof counter
 expose(counter)
 ```
 
-### Error handling
+### TypeScript workers
 
-Works fully transparent - the promise in the master code's call will be rejected with the error thrown in the worker, also yielding the worker error's stack trace.
+**In node.js without webpack** you can spawn `*.ts` workers out-of-the-box without prior transpiling if <a href="https://github.com/TypeStrong/ts-node" rel="nofollow">ts-node</a> is installed.
 
-```js
-// master.js
-import { spawn, Thread, Worker } from "threads"
+If the path passed to `new Worker()` resolves to a `*.ts` file, threads.js will check if `ts-node` is available. If so, it will create an in-memory module that wraps the actual worker module and initializes `ts-node` before running the worker code.
 
-const counter = await spawn(new Worker("./workers/counter"))
+In case `ts-node` is not available, `new Worker()` will attempt to load the same file, but with a `*.js` extension. It is then in your hands to transpile the worker module before running the code.
 
-try {
-  await counter.increment()
-  await counter.increment()
-  await counter.decrement()
-
-  console.log(`Counter is now at ${await counter.getCount()}`)
-} catch (error) {
-  console.error("Counter thread errored:", error)
-} finally {
-  await Thread.terminate(counter)
-}
-```
+**When building your app with webpack**, the module path will automatically be replaced with the path of the worker's resulting bundle file.
 
 ## Observables
 

--- a/package.json
+++ b/package.json
@@ -6,7 +6,6 @@
   "scripts": {
     "prebuild": "rimraf dist/",
     "build": "tsc",
-    "pretest": "babel --plugins @babel/plugin-transform-typescript --plugins @babel/plugin-transform-modules-commonjs ./test/workers -d ./test/workers -x .ts",
     "test": "run-s test:ava test:puppeteer:basic test:puppeteer:webpack",
     "test:ava": "cross-env TS_NODE_FILES=true ava",
     "test:puppeteer:basic": "puppet-run --plugin=mocha --bundle=./test/workers/:/workers/ ./test/*.chromium*.ts",

--- a/src/master/implementation.node.ts
+++ b/src/master/implementation.node.ts
@@ -118,7 +118,7 @@ function initTinyWorker(): typeof WorkerImplementation {
         : resolveScriptPath(scriptPath)
 
       if (resolvedScriptPath.match(/\.tsx?$/i) && detectTsNode()) {
-        super(new Function(createTsNodeModule(resolvedScriptPath)), [], { esm: true })
+        super(new Function(createTsNodeModule(resolveScriptPath(scriptPath))), [], { esm: true })
       } else {
         super(resolvedScriptPath, [], { esm: true })
       }

--- a/src/master/implementation.node.ts
+++ b/src/master/implementation.node.ts
@@ -1,4 +1,4 @@
-// tslint:disable no-eval max-classes-per-file
+// tslint:disable function-constructor no-eval max-classes-per-file
 
 import getCallsites, { CallSite } from "callsites"
 import EventEmitter from "events"
@@ -10,7 +10,31 @@ declare const __non_webpack_require__: typeof require
 
 type WorkerEventName = "error" | "message"
 
+let tsNodeAvailable: boolean | undefined
 const defaultPoolSize = cpus().length
+
+function detectTsNode() {
+  if (typeof __non_webpack_require__ === "function") {
+    // Webpack build: => No ts-node required or possible
+    return false
+  }
+  if (tsNodeAvailable) {
+    return tsNodeAvailable
+  }
+
+  try {
+    require.resolve("ts-node")
+    tsNodeAvailable = true
+  } catch (error) {
+    if (error && error.code === "MODULE_NOT_FOUND") {
+      tsNodeAvailable = false
+    } else {
+      // Re-throw
+      throw error
+    }
+  }
+  return tsNodeAvailable
+}
 
 function createTsNodeModule(scriptPath: string) {
   const content = `
@@ -53,7 +77,7 @@ function initWorkerThreadsWorker(): typeof WorkerImplementation {
     constructor(scriptPath: string) {
       const resolvedScriptPath = resolveScriptPath(scriptPath)
 
-      if (resolvedScriptPath.match(/\.tsx?$/)) {
+      if (resolvedScriptPath.match(/\.tsx?$/i) && detectTsNode()) {
         super(createTsNodeModule(resolvedScriptPath), { eval: true })
       } else {
         super(resolvedScriptPath)
@@ -93,7 +117,7 @@ function initTinyWorker(): typeof WorkerImplementation {
         ? `file:///${resolveScriptPath(scriptPath).replace(/\\/g, "/")}`
         : resolveScriptPath(scriptPath)
 
-      if (resolvedScriptPath.match(/\.tsx?$/)) {
+      if (resolvedScriptPath.match(/\.tsx?$/i) && detectTsNode()) {
         super(new Function(createTsNodeModule(resolvedScriptPath)), [], { esm: true })
       } else {
         super(resolvedScriptPath, [], { esm: true })

--- a/src/master/implementation.node.ts
+++ b/src/master/implementation.node.ts
@@ -12,6 +12,14 @@ type WorkerEventName = "error" | "message"
 
 const defaultPoolSize = cpus().length
 
+function createTsNodeModule(scriptPath: string) {
+  const content = `
+    require("ts-node/register/transpile-only");
+    require(${JSON.stringify(scriptPath)});
+  `
+  return content
+}
+
 function rebaseScriptPath(scriptPath: string, ignoreRegex: RegExp) {
   const parentCallSite = getCallsites().find((callsite: CallSite) => {
     const filename = callsite.getFileName()
@@ -21,7 +29,7 @@ function rebaseScriptPath(scriptPath: string, ignoreRegex: RegExp) {
   const callerPath = parentCallSite ? parentCallSite.getFileName() : null
   const rebasedScriptPath = callerPath ? path.join(path.dirname(callerPath), scriptPath) : scriptPath
 
-  return rebasedScriptPath.replace(/\.ts$/, ".js")
+  return rebasedScriptPath
 }
 
 function resolveScriptPath(scriptPath: string) {
@@ -43,7 +51,14 @@ function initWorkerThreadsWorker(): typeof WorkerImplementation {
     private mappedEventListeners: WeakMap<EventListener, EventListener>
 
     constructor(scriptPath: string) {
-      super(resolveScriptPath(scriptPath))
+      const resolvedScriptPath = resolveScriptPath(scriptPath)
+
+      if (resolvedScriptPath.match(/\.tsx?$/)) {
+        super(createTsNodeModule(resolvedScriptPath), { eval: true })
+      } else {
+        super(resolvedScriptPath)
+      }
+
       this.mappedEventListeners = new WeakMap()
     }
 
@@ -78,7 +93,12 @@ function initTinyWorker(): typeof WorkerImplementation {
         ? `file:///${resolveScriptPath(scriptPath).replace(/\\/g, "/")}`
         : resolveScriptPath(scriptPath)
 
-      super(resolvedScriptPath, [], { esm: true })
+      if (resolvedScriptPath.match(/\.tsx?$/)) {
+        super(new Function(createTsNodeModule(resolvedScriptPath)), [], { esm: true })
+      } else {
+        super(resolvedScriptPath, [], { esm: true })
+      }
+
       allWorkers.push(this)
 
       this.emitter = new EventEmitter()

--- a/test/workers/arraybuffer-xor.ts
+++ b/test/workers/arraybuffer-xor.ts
@@ -1,4 +1,3 @@
-import "ts-node/register"
 import { expose, Transfer } from "../../src/worker"
 
 expose(function xor(buffer: ArrayBuffer, value: number) {

--- a/test/workers/count-to-five.ts
+++ b/test/workers/count-to-five.ts
@@ -1,4 +1,3 @@
-import "ts-node/register"
 import Observable from "zen-observable"
 import { expose } from "../../src/worker"
 

--- a/test/workers/counter.ts
+++ b/test/workers/counter.ts
@@ -1,4 +1,3 @@
-import "ts-node/register"
 import { expose } from "../../src/worker"
 
 let currentCount = 0

--- a/test/workers/faulty-function.ts
+++ b/test/workers/faulty-function.ts
@@ -1,4 +1,3 @@
-import "ts-node/register"
 import { expose } from "../../src/worker"
 
 expose(function fail() {

--- a/test/workers/hello-world.ts
+++ b/test/workers/hello-world.ts
@@ -1,4 +1,3 @@
-import "ts-node/register"
 import { expose } from "../../src/worker"
 
 expose(function helloWorld() {

--- a/test/workers/increment.ts
+++ b/test/workers/increment.ts
@@ -1,4 +1,3 @@
-import "ts-node/register"
 import { expose } from "../../src/worker"
 
 let counter = 0

--- a/test/workers/minmax.ts
+++ b/test/workers/minmax.ts
@@ -1,4 +1,3 @@
-import "ts-node/register"
 import { Observable, Subject } from "../../src/observable"
 import { expose } from "../../src/worker"
 

--- a/test/workers/top-level-throw.ts
+++ b/test/workers/top-level-throw.ts
@@ -1,4 +1,3 @@
-import "ts-node/register"
 import { expose } from "../../src/worker"
 
 throw Error("Top-level worker error")

